homeassistant/components/sense/sensor.py

```diff
@@ -273,6 +277,27 @@ def __init__(
         self._sensor_type = sensor_type
         self._variant_id = variant_id
         self._had_any_update = False
+        self._last_update = None
+        if self._sensor_type == "DAY":
+            self._last_update = dt_util.now().day
+            self._attr_last_reset = dt_util.now().replace(
+                hour=0, minute=0, second=0, microsecond=0
+            )
+        elif self._sensor_type == "WEEK":
+            self._last_update = dt_util.now().day
+            self._attr_last_reset = (dt_util.now() - timedelta(days=7)).replace(
+                hour=0, minute=0, second=0, microsecond=0
+            )
+        elif self._sensor_type == "MONTH":
+            self._last_update = dt_util.now().month
+            self._attr_last_reset = dt_util.now().replace(
+                day=1, hour=0, minute=0, second=0, microsecond=0
+            )
+        elif self._sensor_type == "YEAR":
+            self._last_update = dt_util.now().year
+            self._attr_last_reset = dt_util.now().replace(
+                month=1, day=1, hour=0, minute=0, second=0, microsecond=0
+            )
```

Comment:
This does not seem right. You need to make sure the first state update of a new cycle bumps `last_reset`, modifying the local_time is unlikely to guaranteee that.

Does the data received from the sense API have a timestamp, or even better, a start time of the period?

Reply:
> OK, so how can we tell when a new cycle has started?
I couldn't see a reliable way other than the date changing

> We should not set it at all, and also make sure the sensor's state is None until we have a valid update from the API with the correct data.

That would mean the sensor would of offline for an entire year.   We know when the value was reset (at the start of the period).  The `_attr_last_reset` could be left as `None` if needed, I don't think there would be any issues with that outside that edge case where someone restarts HA right around midnight

Reply:
> I don't think there would be any issues with that outside that edge case where someone restarts HA right around midnight

I think there's a big risk of running into issues due to scheduled daily restarts and the like.
Are you sure there's nothing in the API response from Sense like a timestamp or something else indicating for which period the data is valid?
If there isn't can you reach out to their support and ask if they have a suggested way to reliably know when a new meter cycle starts?

Reply:
My understanding from the code is HA calls this every 5 minutes:
https://github.com/scottbonline/sense/blob/a5d913515e1f83192cbff1bf2a52b1456e764c0f/sense_energy/senseable.py#L79-L81

This means Home Assistant does in fact specify a starting time for the request, and since we do that we should know when `last_reset` should be updated without guessing?

Reply:
The sense library will fetch data from the API every 5 minutes and sends a datetime to the sense server for when it wants data from.  The Sense server isn't that specific about it, so sending 2022-01-10.01:00 in the request will provide the same daily data as 2022-01-10.16:00, I guess it truncates everything after the day.  

We could have it pass the start of the day always ( e.g. 2022-01-10.00:00), then the reset could be kept as that value, but that's basically the same as this is doing

Reply:
The problem with the current code is that if the clocks in the sense server is not well aligned with the clocks on the device running Home Assistant, or if there is a significant delay in the response, an update around midnight (or whenever the sensor cycles) may be interpreted as an update for "tomorrow" by Home Assistant whereas the sense server really sent the last update from "yesterday". This will cause the `last_reset` bump to happen for the wrong state update, which breaks the statistics.

We need to make sure we ask for the correct data from the sense server, and that we remember what data we asked for when we parsed the reply 👍

Reply:
I wouldn't worry about clocks drifting since that's going to be a problem no matter what solution is used due to the way HA records the data.

I guess if it sent a request at 23:59:59 and received the result at 00:00:01 it could cause a problem with a big spike.  We could store the time that we request data from, but it doesn't look that'll be easily accessible while using the `DataUpdateCoordinator` so it'll have to be a new field for the component or something.  With the datetime of the request stored, we can then use the same code determine if it's reset